src/PerfView/StackViewer/CallTreeView.cs

```diff
@@ -324,49 +324,63 @@ public bool IsExpanded
-                {
-                    ValidateTree();
+                    ExpandNode();
+                else
+                    CollapseNode();
+            }
+        }
 
-                    // We are trying to expand the treeNode, add the children after me. 
-                    var children = MakeChildren();
-                    m_treeView.m_flattenedTree.InsertRange(MyIndex + 1, children);
-                    m_isExpanded = true;
+        private void ExpandNode(bool selectExpandedNode = true)
+        {
+            if (m_isExpanded)
+                return;
 
-                    ValidateTree();
-                    // Auto expand nodes that have only one real child.  (Don't do this for graph nodes as it may not terminate.  
-                    if (children.Count == 1 && !Data.IsGraphNode)
-                    {
-                        var onlyChild = children[0];
-                        if (onlyChild.HasChildren)
-                            onlyChild.IsExpanded = true;
-                    }
-                    else
-                        m_treeView.m_perfGrid.Select(this);         // We want expanding the node to select the node
-                }
-                else
-                {
-                    ValidateTree();
+            ValidateTree();
 
-                    int firstChild = MyIndex + 1;
-                    int lastChild = firstChild;
-                    int myDepth = m_depth;
-                    while (lastChild < m_treeView.m_flattenedTree.Count)
-                    {
-                        if (m_treeView.m_flattenedTree[lastChild].m_depth <= myDepth)
-                            break;
-                        lastChild++;
-                    }
+            // We are trying to expand the treeNode, add the children after me. 
+            var children = MakeChildren();
+            m_treeView.m_flattenedTree.InsertRange(MyIndex + 1, children);
+            m_isExpanded = true;
 
-                    m_treeView.m_flattenedTree.RemoveRange(firstChild, lastChild - firstChild);
-                    m_isExpanded = false;
+            ValidateTree();
+            // Auto expand nodes that have only one real child.  (Don't do this for graph nodes as it may not terminate.  
+            if (children.Count == 1 && !Data.IsGraphNode)
+            {
+                var onlyChild = children[0];
+                if (onlyChild.HasChildren)
+                    onlyChild.ExpandNode(selectExpandedNode);
+            }
+            else if (selectExpandedNode)
+                m_treeView.m_perfGrid.Select(this);         // We want expanding the node to select the node
 
-                    // set the selected node to my caller (if available) or myself if there is none.  
-                    m_treeView.Select(Data.Caller ?? Data);  
-                }
-                ValidateTree();
+            ValidateTree();
+        }
+
+        private void CollapseNode()
+        {
+            ValidateTree();
```

Comment:
Can you put a check (!m_isExpanded) return  at the start to be symmetric with your ExpandNode routine.    Seems safer now that others might call this from locations where the node happened to already be collapsed.